apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/ids/GlobalIdGenerator.java

```diff
@@ -74,11 +73,11 @@ private long timestamp() {
 
             if (currentTimeMillis < lastTimestamp) {
                 // Just for considering time-shift-back by Ops or OS. @hanahmily 's suggestion.
-                if (runRandomTimestamp != currentTimeMillis) {
-                    lastRandomValue = ThreadLocalRandom.current().nextInt();
-                    runRandomTimestamp = currentTimeMillis;
+                if (lastShiftTimestamp != currentTimeMillis) {
+                    lastShiftValue++;
```

Comment:
Again, this has a race condition. And no guarantee in concurrency

Reply:
![image](https://user-images.githubusercontent.com/16626084/114403596-85344000-9bd7-11eb-825d-10319b0dd8ef.png)

Reply:
So? ThreadLocal is just holding a value? Then it could be as same as `N+1` in different threads.

Reply:
对于每一次调用`GlobalIdGenerator.generate()`，都是获取当前线程的`ThreadLocal `对应的`IDContext`，也就说每个线程获取的`IDContext`都是自己独有的，不会存在并发问题。
`IDContext`不存在并发问题的话，显然，其中的`lastShiftValue`也不会。

For each call to `GlobalIdGenerator.generate()`, the `IDContext` corresponding to the `ThreadLocal` of the current thread is obtained. In other words, the `IDContext` obtained by each thread is unique to itself, and there will be no concurrency problems. .
If there is no concurrency problem with `IDContext`, obviously, the `lastShiftValue` in it will not.

Reply:
我的理解没有问题吧？我看下面的的`nextThreadSeq++`也是没有问题
There is no problem with my understanding, right? I see the `nextThreadSeq++` below is also no problem
![image](https://user-images.githubusercontent.com/16626084/114813287-5fc85180-9de4-11eb-8d79-e23952f90709.png)